shaders/khr_ray_tracing_extended/closesthit.rchit

```diff
@@ -0,0 +1,168 @@
+in     int   gl_PrimitiveID;
+in     int   gl_InstanceID;
+in     int   gl_InstanceCustomIndexEXT;
+in     int   gl_GeometryIndexEXT;
+*/
+
+struct Vertex
+{
+  vec3 pt;
+  vec3 normal;
+  vec2 coordinate;
+};
+
+Vertex getVertex(uint vertexOffset, uint index, bool is_static)
+{
+  uint base_index = 2 * (vertexOffset + index);
+  vec4 A = is_static ? vertex_buffer.data[base_index] : dynamic_vertex_buffer.data[base_index];
+  vec4 B = is_static ? vertex_buffer.data[base_index + 1] : dynamic_vertex_buffer.data[base_index + 1];
+
+  Vertex v;
+  v.pt = A.xyz;
+  v.normal = vec3(A.w, B.x, B.y);
+  v.coordinate = vec2(B.z, B.w);
+  return v;
+}
+
+uvec3 getIndices(uint triangle_offset, uint primitive_id, bool is_static)
+{
+    uint base_index = 3 * (triangle_offset + primitive_id);
+    uint index0 = is_static ? index_buffer.indices[base_index] : dynamic_index_buffer.indices[base_index];
+    uint index1 = is_static ? index_buffer.indices[base_index + 1] : dynamic_index_buffer.indices[base_index + 1];
+    uint index2 = is_static ? index_buffer.indices[base_index + 2] : dynamic_index_buffer.indices[base_index + 2];
+
+    return uvec3(index0, index1, index2);
+}
+
+void handleDraw()
+{
+    uint index = gl_InstanceCustomIndexEXT;
+
+    uint vertexOffset = data_map.indices[4 * index];
+    uint triangleOffset = data_map.indices[4*index + 1];
+    uint imageOffset = data_map.indices[4 * index + 2];
+    uint objectType = data_map.indices[4 * index + 3];
+    bool is_static = objectType != 1;
+
+    uvec3 indices = getIndices(triangleOffset, gl_PrimitiveID, is_static);
+    Vertex A = getVertex(vertexOffset, indices.x, is_static), B = getVertex(vertexOffset, indices.y, is_static), C = getVertex(vertexOffset, indices.z, is_static);
+
+    // interpolate and obtain world point
+    const vec3 barycentricCoords = vec3(1.0f - attribs.x - attribs.y, attribs.x, attribs.y);
+    float alpha = barycentricCoords.x, beta = barycentricCoords.y, gamma = barycentricCoords.z;
+    vec3 pt = alpha * A.pt + beta * B.pt + gamma * C.pt;
+    mat4x3 transform = gl_WorldToObjectEXT;
+    vec3 worldPt =  gl_WorldRayOriginEXT + gl_HitTEXT * gl_WorldRayDirectionEXT;//transform * vec4(pt, 0) + vec3(transform[3][0], transform[3][1], transform[3][2]);
+    vec3 normal = normalize(alpha * A.normal + beta * B.normal + gamma * C.normal);
+    vec3 worldNormal = normalize(cross(B.pt - A.pt, C.pt - A.pt));
+
+    vec2 texcoord = alpha * A.coordinate + beta * B.coordinate + gamma * C.coordinate;
+
+    hitValue.intersection = vec4(worldPt.xyz, objectType);
+    hitValue.normal = vec4(worldNormal.xyz, gl_HitTEXT);
+    if (render_mode == 4) { // global xyz
+      hitValue.color = vec4(heatmap(worldPt.x, -10, 10), 1);
+      return;
+    }
+    if ((objectType == 0 || objectType == 2)){
+      if (imageOffset >= 26){
+        return; // this shouldn't happen
+      }
+      // obtain texture coordinate
+      vec4 tex_value = texture(textures[imageOffset], texcoord);
```

Comment:
Should `imageOffset` be wrapped by `nonuniformEXT` here? Also, are derivatives valid-- or should this be using `textureLod`?

Reply:
derivatives I believe are valid, or I at least have no reason to think otherwise.  This isn't using mipmaps for the main reason that I didn't want to make the sample too complicated.  Meaning if I add mipmaps, then why not add don't update things not in view?  or update the particle effect TLAS to use less particles if the LOD is far away etc.

I agree it should be nonuniformEXT.  I'll submit another commit with it corrected, thanks!

Reply:
Cheers :) If there's no mips, I would maybe argue using `textureLod(..., 0)` would be better -- as there's a risk people will copy this code as it _is_ sample. I somewhat doubt derives are valid TBH, so my guess likely only works properly as long as there's only 1 mip.

Reply:
That's actually a very good point.  I'll switch to textureLod and leave a comment that spells out reasoning.

Thanks!